src/main/groovy/org/jenkinsci/gradle/plugins/jpi/LicenseTask.groovy

```diff
@@ -0,0 +1,94 @@
+import org.gradle.api.artifacts.ResolvedArtifact
+import org.gradle.api.tasks.OutputDirectory
+import org.gradle.api.tasks.TaskAction
+import org.gradle.internal.component.local.model.DefaultProjectComponentIdentifier
+
+class LicenseTask extends DefaultTask {
+    static final String TASK_NAME = 'generateLicenseInfo'
+
+    private Set<Configuration> configurations
+
+    Set<Configuration> providedConfigurations
+
+    @OutputDirectory
+    File outputDirectory
+
+    void setConfigurations(Set<Configuration> configurations) {
+        this.configurations = configurations
+        this.inputs.files(configurations*.incoming.files)
+    }
+
+    @TaskAction
+    void generateLicenseInfo() {
+        JpiExtension jpiExtension = project.extensions.getByType(JpiExtension)
+        Set<ResolvedArtifact> pomArtifacts = collectPomArtifacts()
+
+        new File(outputDirectory, 'licenses.xml').withWriter { Writer writer ->
+            MarkupBuilder xmlMarkup = new MarkupBuilder(writer)
+
+            xmlMarkup.'l:dependencies'(
+                    'xmlns:l': 'licenses', version: project.version, artifactId: project.name, groupId: project.group,
+            ) {
+                'l:dependency'(
+                        version: project.version, artifactId: project.name, groupId: project.group,
+                        name: jpiExtension.displayName, url: jpiExtension.url,
+                ) {
+                    'l:description'(project.description)
+                    jpiExtension.licenses.each { license ->
+                        'l:license'(url: license.url, name: license.name)
+                    }
+                }
+
+                pomArtifacts.each { ResolvedArtifact pomArtifact ->
+                    Node pom = new XmlParser().parse(pomArtifact.file)
+
+                    ModuleVersionIdentifier gav = pomArtifact.moduleVersion.id
+                    String name = pom[QName.valueOf('name')].text()
+                    String description = pom[QName.valueOf('description')].text()
+                    String url = pom[QName.valueOf('url')].text()
+                    NodeList licenses = pom[QName.valueOf('licenses')]
+
+                    'l:dependency'(
+                            version: gav.version, artifactId: gav.name, groupId: gav.group, name: name, url: url,
+                    ) {
+                        'l:description'(description)
+                        licenses[QName.valueOf('license')].each { Node license ->
+                            String licenseUrl = license[QName.valueOf('url')].text()
+                            String licenseName = license[QName.valueOf('name')].text()
+                            'l:license'(url: licenseUrl, name: licenseName)
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    private Set<ResolvedArtifact> collectPomArtifacts() {
+        Configuration pomConfiguration = project.configurations.create('poms')
```

Comment:
Shouldn't the plugin add this configuration? Or we should at least check if the configuration already exists and then re-use it.

Reply:
I would rather not expose this as it is an implementation detail. But avoiding a collision with an existing configuration makes sense.

Reply:
If you don't want to expose the configuration, why don't you use [`detachedConfiguration`](https://github.com/gradle/gradle/blob/209ed1fb85fddbe66dd661b0a638ddb958bed1fe/subprojects/core-api/src/main/java/org/gradle/api/artifacts/ConfigurationContainer.java#L114) instead of `create`?

Reply:
Mainly because I didn't know about detached configurations.